Adds mock implementation for MD3 scan commands.

mockoocy opened the pull request:
This includes mocks of `startScanEx` and `startScan4DEx` commands of MD3.
They are mocked by moving the motoros (in a similar fashion to already existing `Exporter._move_motor` function) and opening/closing fast shutter.

mockoocy committed:
Adds mock implementation for MD3 scan commands.

This includes mocks of `startScanEx` and `startScan4DEx` commands of
MD3.
They are mocked by moving the motoros (in a similar fashion to already
existing `Exporter._move_motor` function) and opening/closing fast
shutter.

fabcor-maxiv reviewed (approved):
Seems fine to me. Better wait for @elmjag's approval...

elmjag commented on b-micromax-md3-pc/exporter.py:
```diff
@@ -444,15 +444,74 @@ def _do_start_raster_scan(
-        return self._add_task("Start 4D-SCAN", 3.2)
+        TASK_DURATION = 3.2 
+        async def start_scan():
+            self.write_attribute("FastShutterIsOpen", True)
```
You need to take into account the `start_angle` parameter as well.

The `startScanEx` command _first_ rotates Omega to `start_angle` and then rotates to `start_angle` + `scan_range` angle.

mockoocy replied:
Made it so that it respects the `start_angle`. Moved it to separate method (as there was a lot of same-y logic regarding moving these motors.

elmjag commented on b-micromax-md3-pc/exporter.py:
```diff
@@ -444,15 +444,74 @@ def _do_start_raster_scan(
-
-    def _do_start_scan_4d_ex(self, *_):
-        return self._add_task("Start 4D-SCAN", 3.2)
+        TASK_DURATION = 3.2 
```
Use the `exposure_time` argument instead. It does specify how long time the scan should be performed.

Technically speaking, it specifies how long time the rotation between `start_angle` and `start_angle` + `scan_range` angles takes. The hole scan command take more time, but we can ignore these details for now.

elmjag commented on b-micromax-md3-pc/exporter.py:
```diff
@@ -444,15 +444,74 @@ def _do_start_raster_scan(
+            for _ in range(MOTOR_STEPS):
+                current_pos = self._attrs[pos_attr].val
+                log(f"INFO: {current_pos=} for {pos_attr=}, will move by {step_size=}")
+                self.write_attribute(pos_attr, current_pos + step_size)
```
The `Omega` value needs to wrap around 0..360 range. See for example how it's done here:

https://github.com/elmjag/micromax/blob/9332db83cca50ba87a0e5027c6310b31fbad24ea/b-micromax-md3-pc/exporter.py#L636

I think it's time to introduce a dedicated function for this.

```
def _wrap_omega_position(new_omega_position: float) -> float:
  """Keep Omega rotation angle within 0..360 range"""
  return new_omega_position % 360.0
```

mockoocy committed:
Now respects start_angle and exposure_time.

More closely mimicks the actual MD3 behaviour, in which omega is first
moved to `start_angle` position, before the actual scan starts.

This also moves logic regarding moving motors to a separate method.

elmjag commented:
Very nice!

I have made some modifications, squashed all commits and [pushed it](https://github.com/elmjag/micromax/commit/b10409da096c1987719b0e93ab4351ca13301ea3).
